test/src/unit-bson.cpp

```diff
@@ -0,0 +1,1115 @@
+of this software and associated  documentation files (the "Software"), to deal
+in the Software  without restriction, including without  limitation the rights
+to  use, copy,  modify, merge,  publish, distribute,  sublicense, and/or  sell
+copies  of  the Software,  and  to  permit persons  to  whom  the Software  is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE  IS PROVIDED "AS  IS", WITHOUT WARRANTY  OF ANY KIND,  EXPRESS OR
+IMPLIED,  INCLUDING BUT  NOT  LIMITED TO  THE  WARRANTIES OF  MERCHANTABILITY,
+FITNESS FOR  A PARTICULAR PURPOSE AND  NONINFRINGEMENT. IN NO EVENT  SHALL THE
+AUTHORS  OR COPYRIGHT  HOLDERS  BE  LIABLE FOR  ANY  CLAIM,  DAMAGES OR  OTHER
+LIABILITY, WHETHER IN AN ACTION OF  CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "catch.hpp"
+
+#include <nlohmann/json.hpp>
+using nlohmann::json;
+#include <fstream>
+
+TEST_CASE("BSON")
+{
+    SECTION("individual values not supported")
+    {
+        SECTION("discarded")
+        {
+            // discarded values are not serialized
+            json j = json::value_t::discarded;
+            const auto result = json::to_bson(j);
+            CHECK(result.empty());
+        }
+
+        SECTION("null")
+        {
+            json j = nullptr;
+            REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+        }
+
+        SECTION("boolean")
+        {
+            SECTION("true")
+            {
+                json j = true;
+                REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+            }
+
+            SECTION("false")
+            {
+                json j = false;
+                REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+            }
+        }
+
+        SECTION("number")
+        {
+            json j = 42;
+            REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+        }
+
+        SECTION("float")
+        {
+            json j = 4.2;
+            REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+        }
+
+        SECTION("string")
+        {
+            json j = "not supported";
+            REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+        }
+
+        SECTION("array")
+        {
+            json j = std::vector<int> {1, 2, 3, 4, 5, 6, 7};
+            REQUIRE_THROWS_AS(json::to_bson(j), json::type_error);
+        }
+    }
+
+    SECTION("keys containing code-point U+0000 cannot be serialized to BSON")
+    {
+        json j =
+        {
+            { std::string("en\0try", 6), true }
+        };
+        REQUIRE_THROWS_AS(json::to_bson(j), json::out_of_range);
+        CHECK_THROWS_WITH(json::to_bson(j), "[json.exception.out_of_range.409] BSON key cannot contain code point U+0000");
```

Comment:
Is it possible to add an offset (i.e., the byte position) to simplify understanding and fixing the error?

Reply:
Good point! I adjusted the code accordingly to report the position within the key-string.